pkg/channel/fanout/fanout_message_handler.go

```diff
@@ -164,17 +166,33 @@ func (f *MessageHandler) dispatch(ctx context.Context, bufferedMessage binding.M
 // makeFanoutRequest sends the request to exactly one subscription. It handles both the `call` and
 // the `sink` portions of the subscription.
 func (f *MessageHandler) makeFanoutRequest(ctx context.Context, message binding.Message, additionalHeaders nethttp.Header, sub eventingduck.SubscriberSpec) error {
+
 	var destination *url.URL
 	if sub.SubscriberURI != nil {
 		destination = sub.SubscriberURI.URL()
 	}
+
 	var reply *url.URL
 	if sub.ReplyURI != nil {
 		reply = sub.ReplyURI.URL()
 	}
-	var deadLetter *url.URL
+
+	var deadLetterURL *url.URL
 	if sub.Delivery != nil && sub.Delivery.DeadLetterSink != nil && sub.Delivery.DeadLetterSink.URI != nil {
-		deadLetter = sub.Delivery.DeadLetterSink.URI.URL()
+		deadLetterURL = sub.Delivery.DeadLetterSink.URI.URL()
 	}
-	return f.dispatcher.DispatchMessage(ctx, message, additionalHeaders, destination, reply, deadLetter)
+
+	retriesConfigs := kncloudevents.NoRetries()
+	if sub.Delivery != nil {
+
+		delivery := &eventingduckv1.DeliverySpec{}
+		_ = sub.Delivery.ConvertTo(ctx, delivery)
+
+		_retriesConfigs, err := kncloudevents.RetryConfigFromDeliverySpec(*delivery)
```

Comment:
This method is expensive on the hot path and its result can be cached. You shouldn't perform the parsing from delivery spec to retries config every time. I suggest you change the internal data structure of `Config` to put the parsed retry config there and who builds the `MessageHandler` takes care of parsing the retries config

Reply:
Each subscription has its own delivery spec, are you suggesting to have a map with a lock to index retries config for reach subscription?

Reply:
Today config is:

```
type Config struct {
	Subscriptions []eventingduck.SubscriberSpec `json:"subscriptions"`
	// AsyncHandler controls whether the Subscriptions are called synchronous or asynchronously.
	// It is expected to be false when used as a sidecar.
	AsyncHandler bool `json:"asyncHandler,omitempty"`
}
```

We might change it to:

```
type SubscriptionConfig struct {
  // Only what we need from duck subscriber spec
  retryConfig *kncloudevents.RetriesConfig
}

type Config struct {
	Subscriptions []SubscriptionConfig
	AsyncHandler bool
}
```

Reply:
Are you sure this is not breaking?

Reply:
See https://github.com/knative/eventing/blob/master/pkg/channel/multichannelfanout/parse.go and https://github.com/knative/eventing/blob/master/pkg/channel/multichannelfanout/config.go

Reply:
What if we have a "parallel" array of type `[]RetriesConfig`? `retries[i] = retriesOf(subscriptions[i])`

Reply:
This is breaking, I'm aware of that. To being less breaking, we might do:

```
type SubscriptionConfig struct {
  eventingduck.SubscriberSpec
  retryConfig *kncloudevents.RetriesConfig
}
```

Reply:
Done: https://github.com/knative/eventing/pull/2932/commits/e029985ffd76e024966088316fc07447ec01f2fe

Reply:
Please, let me know if I can go ahead with this: https://github.com/knative/eventing/pull/2932#discussion_r457111554.

Reply:
ok now it's good and we didn't break anything :smile: go ahead with the signature of MessageDispatcher